Populate OnlineClient.Properties from HttpContext.Query

fncap opened the pull request:
Resolves #5612

maliming commented:
From https://support.aspnetzero.com/QA/Questions/9078

maliming commented:
hi @fncap 

We can consider adding an `IOnlineClientInfoProvider `interface to create `OnlineClient`. Just like the `IClientInfoProvider `interface.

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/37a46288e96a12b5a5318b84fd7bd659b61dc7f4/src/Abp.AspNetCore.SignalR/AspNetCore/SignalR/Hubs/OnlineClientHubBase.cs#L71

This way you can replace provider through dependency injection.

The default implementation we keep the current implementation.
```c#
new OnlineClient(	            
	Context.ConnectionId,	
	GetIpAddressOfClient(),	
	Context.GetTenantId(),	
	Context.GetUserIdOrNull()
);
```

fncap commented:
Hi @maliming 

thanks for your suggestion! Inspecting Abp code, I found two different implementations of `IClientInfoProvider`. Should I implements both for `IOnlineClientProvider`?
I don't know well where put thoose implementations (in which folder I mean), but I'll do it asap using this PR so you can review it and suggest corrections if needed!

maliming commented:
They are net core and net framework mvc5.

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/35cb7f6be1b9a8a17f64f83f82567c09b2c5bf07/src/Abp.AspNetCore/AspNetCore/Mvc/Auditing/HttpContextClientInfoProvider.cs#L9

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/b8fb979b9f586d9aa440fe2c7295e3a8c81307d2/src/Abp.Web/Auditing/WebClientInfoProvider.cs#L10


You need to implement them at the same time.

fncap commented:
Thanks! I'll do asap!

fncap commented:
Just for clarification: are you intend to replace 

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/35cb7f6be1b9a8a17f64f83f82567c09b2c5bf07/src/Abp.AspNetCore.SignalR/AspNetCore/SignalR/Hubs/OnlineClientHubBase.cs#L57-L65

and

https://github.com/aspnetboilerplate/aspnetboilerplate/blob/35cb7f6be1b9a8a17f64f83f82567c09b2c5bf07/src/Abp.Web.SignalR/Web/SignalR/Hubs/AbpCommonHub.cs#L82-L90

with something like 

```c#
private IOnlineClient CreateClientForCurrentConnection()
{
   return _onlineClientInfoProvider.Create(Context);
}
```

In that case, both implementations of `IOnlineClientInfoProvider` should reside inside - respectively - Abp.AspNetCore.SignalR and Abp.Web.SignalR modules due to their dependencies with `Context` argument passed, which is of `Microsoft.AspNet.SignalR.Hubs.HubCallerContext` type!

If this is correct, then it could be possible to move `IClientInfoProvider` from both Hub implementations directly inside `IOnlineClientInfoProvider`!

maliming commented:
However, their parameter types are different, and using `object `as a parameter is not a good design. We can consider making this interface exist in net core and net framework respectively.

fncap commented:
Hi @maliming,
in the last commit, I implemented `IOnlineClientInfoProvider` accordingly to our thread.
BTW I removed code portion responsible for populating `OnlineClient.Properties` from QueryString (as per my suggestion). 
Is this enhancement still valid so I can re-implement it's code directly inside `IOnlineClientInfoProvider`?

maliming commented on src/Abp.Web.SignalR/RealTime/OnlineClientInfoProvider.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Abp.RealTime
+{
+    public class OnlineClientInfoProvider : AbpServiceBase, IOnlineClientInfoProvider
```
I don't think it needs to inherit from `AbpServiceBase`. Currently it just needs `IAbpSession`.

fncap replied:
In the 0d7540e9615f2f74f01cb780f7b90acc24a28d31 I removed unwanted `AbpServiceBase` inheritance! Thanks!

maliming commented:
hi

By default we don't use `QueryString` to populate `OnlineClient.Properties`. You can replace the implementation of `IOnlineClientInfoProvider `in your application.
